[quant] Add Graph Mode Passes to quantize EmbeddingBag operators

supriyar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#41612 [quant] Add Graph Mode Passes to quantize EmbeddingBag operators**

Summary:
This change adds preliminary support to quantize the EmbeddingBag operators. We currently support 4-bit and 8-bit quantization+packing of the weights.

To quantize these operators, specify the operator name in the `custom_op_name` field of the NoopObserver. Based on the op name (4bit or 8bit) we call the corresponding quantization functions.
Refer to the testplan for how to invoke the qconfig for the embedding_bag ops.

Future versions of this will support 4-bit and 2-bit qtensors with native support to observe and quantize it.

NB - This version assumes that the weights in the EmbeddingBag Module reside on the same device.

Test Plan:
python test/test_quantization.py TestQuantizeDynamicJitOps.test_embedding_bag

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22609342](https://our.internmc.facebook.com/intern/diff/D22609342)

supriyar committed:
[quant] Add Graph Mode Passes to quantize EmbeddingBag operators
Summary:
This change adds preliminary support to quantize the EmbeddingBag operators. We currently support 4-bit and 8-bit quantization+packing of the weights.

To quantize these operators, specify the operator name in the `custom_op_name` field of the NoopObserver. Based on the op name (4bit or 8bit) we call the corresponding quantization functions.
Refer to the testplan for how to invoke the qconfig for the embedding_bag ops.

Future versions of this will support 4-bit and 2-bit qtensors with native support to observe and quantize it.

NB - This version assumes that the weights in the EmbeddingBag Module reside on the same device.

Test Plan:
python test/test_quantization.py TestQuantizeDynamicJitOps.test_embedding_bag

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "[quant] Add Graph Mode Passes to quantize EmbeddingBag operators"


Summary:
This change adds preliminary support to quantize the EmbeddingBag operators. We currently support 4-bit and 8-bit quantization+packing of the weights.

To quantize these operators, specify the operator name in the `custom_op_name` field of the NoopObserver. Based on the op name (4bit or 8bit) we call the corresponding quantization functions.
Refer to the testplan for how to invoke the qconfig for the embedding_bag ops.

Future versions of this will support 4-bit and 2-bit qtensors with native support to observe and quantize it.

NB - This version assumes that the weights in the EmbeddingBag Module reside on the same device.

Test Plan:
python test/test_quantization.py TestQuantizeDynamicJitOps.test_embedding_bag

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22609342](https://our.internmc.facebook.com/intern/diff/D22609342)

[ghstack-poisoned]

supriyar committed:
Update on "[quant] Add Graph Mode Passes to quantize EmbeddingBag operators"


Summary:
This change adds preliminary support to quantize the EmbeddingBag operators. We currently support 4-bit and 8-bit quantization+packing of the weights.

To quantize these operators, specify the operator name in the `custom_op_name` field of the NoopObserver. Based on the op name (4bit or 8bit) we call the corresponding quantization functions.
Refer to the testplan for how to invoke the qconfig for the embedding_bag ops.

Future versions of this will support 4-bit and 2-bit qtensors with native support to observe and quantize it.

NB - This version assumes that the weights in the EmbeddingBag Module reside on the same device.

Test Plan:
python test/test_quantization.py TestQuantizeDynamicJitOps.test_embedding_bag

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22609342](https://our.internmc.facebook.com/intern/diff/D22609342)

[ghstack-poisoned]

vkuzo commented on test/quantization/test_quantize_jit.py:
```diff
@@ -2813,6 +2814,43 @@ def forward(self, x):
+                e2 = self.embedding2(indices2, offsets2)
+                return e1, e2
+
+        weights = torch.from_numpy((np.random.random_sample((10, 12)) + 1).astype(np.float32))
```
optional: torch.randn to remove numpy dependency?

vkuzo commented on torch/csrc/jit/passes/quantization/helper.h:
```diff
@@ -99,6 +99,11 @@ TORCH_API bool useQuantizable(const Use& use, QuantType quant_type);
 // Given a CallFunction node, extract the graph of the called function
 TORCH_API std::shared_ptr<Graph> getCallFunctionGraph(Node* n);
 
+TORCH_API bool matchCallFuncToUse(
```
optional nit: docblock?

jerryzh168 replied:
nit: I think we can remove TORCH_API actually, it's a mistake in the beginning to add TORCH_API for all these helper functions..

vkuzo commented on torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp:
```diff
@@ -224,6 +241,49 @@ bool isNoopObserver(Value* observer) {
+      isNoopObserver(observer);
+}
+
+std::string getEmbeddingBagObsName(script::Module& module, Node* n) {
```
should this return `c10::optional`, like `findObserverName` does?

vkuzo commented on torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp:
```diff
@@ -224,6 +241,49 @@ bool isNoopObserver(Value* observer) {
+  return "";
+}
+
+std::tuple<Node*, Node*> insertDefaultObserverNodes(
```
nit: brief docblock?

vkuzo commented on test/quantization/test_quantize_jit.py:
```diff
@@ -2813,6 +2814,43 @@ def forward(self, x):
+        offsets = torch.tensor([0, 19, 20, 28, 28, 32])
+
+        from torch.quantization import QConfigDynamic, NoopObserver
+        int4_dynamic_qconfig = QConfigDynamic(activation=NoopObserver.with_args(custom_op_name="embedding_bag_4bit"),
```
makes sense that `custom_op_name` is a temporary solution.  Do we have thoughts on what to replace it with eventually /  why not now?  Would it be `EmbeddingBag{8|4|2}BitObserver` / something else?

jerryzh168 replied:
what's the longer term solution? are we planning to add `torch.qint4`, `torch.qint2` etc.?

supriyar replied:
yes, long term this will be replaced with observer that can support `torch.qint4` and `torch.qint2`

vkuzo commented on torch/csrc/jit/passes/quantization/helper.h:
```diff
@@ -99,6 +99,11 @@ TORCH_API bool useQuantizable(const Use& use, QuantType quant_type);
+TORCH_API bool matchCallFuncToUse(
+    const Use& use,
+    const std::string& func_name,
+    c10::optional<int> n);
```
what does n mean here, maybe we can name it more descriptively?

vkuzo reviewed (approved):
lgtm, accepting to unblock.  Feel free to wait for @jerryzh168 if a deeper review on the JIT pass is needed.

jerryzh168 commented on torch/csrc/jit/passes/quantization/helper.cpp:
```diff
@@ -257,7 +258,7 @@ bool isWeight(Value* v) {
       AtenFuncArgs(
           {{"conv1d", 1}, {"conv2d", 1}, {"conv3d", 1}, {"linear", 1}}),
-      CallFuncArgs({{"linear", 2}}));
+      CallFuncArgs({{"linear", 2}, {"embedding_bag", 2}}));
```
I feel we should add some docs for these configurations to make it more clear

supriyar replied:
Do you mean adding the op signature here for every op?

jerryzh168 replied:
yeah

jerryzh168 commented on torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp:
```diff
@@ -224,6 +241,49 @@ bool isNoopObserver(Value* observer) {
+  auto observer_module = module.attr(findObserverName(v).value()).toModule();
+  if (observer_module.hasattr("custom_op")) {
+    auto op_name = observer_module.attr("custom_op").toStringRef();
+    return isNoopObserver(observer) ? op_name : "";
```
follow up PR: since NoopObserver is special, probably better to add a "_" prefix to reserve this for internal use.

jerryzh168 commented on torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp:
```diff
@@ -224,6 +241,49 @@ bool isNoopObserver(Value* observer) {
+    Node* observer,
+    const std::vector<std::string>& qparam_names,
+    const std::string& quantize_func) {
+  // Else branch is executed for dynamic weight observers and all observers
```
need to update comment

jerryzh168 commented on torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp:
```diff
@@ -245,50 +305,39 @@ void insertQuantizationOps(
-      inputs.push_back(g->insertGetAttr(self, qparam_name));
+
+  // Special case for embedding bag operators
+  if (quant_type == QuantType::DYNAMIC &&
```
can we make this branch one of the inner branch of the DYNAMIC branch starting in L318?

jerryzh168 replied:
wait, what is this doing?

supriyar replied:
`insertEmbeddingBagQuantOps` below is run only for the weight tensors. For the indices input we still have an observer attached which we need to disregard. This part of code replaces that observer output with the original value so that the observer can be deleted in cleanup step. I'll update comment to make it more descriptive.

jerryzh168 replied:
why can't we skip inserting observer for input? I remember we implemented that support before for batchnorm?

jerryzh168 replied:
https://codebrowser.bddppq.com/pytorch/pytorch/torch/csrc/jit/passes/quantization/helper.cpp.html#torch::jit::_observe_inputs_call_func

supriyar replied:
I think we spoke offline about this as well but this case is special so we need to observe both. If we don't I run into 
```
RuntimeError: branches for `if` should return values that are observed consistently, if node:%input.22 : Tensor, %weight : Tensor = prim::If(%19) # /home/supriyar/local/miniconda3/envs/pytorch/lib/python3.7/site-packages/torch/nn/functional.py:1908:4
```

jerryzh168 replied:
Oh I see, then could you add a comment for this?

jerryzh168 commented on torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp:
```diff
@@ -1158,6 +1293,7 @@ void InsertQuantDeQuantHelper::runWeightObserver(
   // weight.
   for (const auto& v : weight_values) {
     extractAndRunWeightObserver(module, self, v);
+    insertEmbeddingBagQuantOps(module, self, v);
```
why does this happens here?

jerryzh168 replied:
currently we insert prepack/unpack ops here: https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/passes/quantization/finalize.cpp#L50

jerryzh168 replied:
if we need the observer info why can't we insert this op at the same place when we insert quant/dequant ops?

supriyar replied:
That's mainly because the logic where we insert quant/dequant ops doesn't get called for every module instance. I added it here to make sure the prepack and quant ops are added for each embeddingBag module instance.

jerryzh168 replied:
the current logic is we insert for each graph, but the graph itself might be shared by multiple module instances, why that does not work?

supriyar replied:
My understanding was that we cannot have a shared graph here as each module instance will need its own set of pack and quant ops inserted. I initially did take this approach but I noticed that in the final graph only one of the embedding modules had the quant ops inserted.

jerryzh168 replied:
you mean only one of the embedding modules has the packing? or only one has quant + packing and the other one is just ignored?

jerryzh168 replied:
I don't see why we can't have a shared graph here, could you show the problem with putting this together with insert quant/dequant in more detail?

supriyar replied:
@jerryzh168 I think it should be fine. I realized my test graph wasn't doing the right thing. Will update to move to insert quant_dequant function.

jerryzh168 commented on torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp:
```diff
@@ -1180,15 +1316,17 @@ void InsertQuantDeQuantHelper::run(
-        const auto& name = pr.first;
-        const auto& qparam = pr.second;
-        module._ivalue()->setAttr(qparam_name_map.at(name), qparam);
+      if (qparam_map.size() > 0) {
```
I think it's better to keep the insert here. breaking this insert means there is something wrong and we'll need to fix that

jerryzh168 replied:
when is the qparam map empty?

supriyar replied:
Noop observer is an example of an observer that doesn't have qparam_map

jerryzh168 replied:
OK, it's probably fine in this case, could you also add a comment for this check?

jerryzh168 commented on torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp:
```diff
@@ -782,6 +831,92 @@ void InsertQuantDeQuantHelper::extractAndRunWeightObserver(
+  }
+  // Insert prepack op
+  Node* prepack = g->create(Symbol::fromQualString(prepack_fn), prepack_inputs);
+  g->insertNode(prepack);
```
could you also add a `WithInsertPoint`? we want to insert before the use node I think.

jerryzh168 replied:
should just be: `WithInsertPoint ins(embedding_bag_loat_op);`

supriyar replied:
We've already added the insert point in insertQuantizationOps at the output of the observer node.

jerryzh168 commented on torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp:
```diff
@@ -782,6 +831,92 @@ void InsertQuantDeQuantHelper::extractAndRunWeightObserver(
+  std::vector<Value*> prepack_inputs = {observer_out};
+  std::vector<Use> uses = observer_out->uses();
+  Node* embedding_bag_float_op;
+  for (const Use& use : uses) {
+    if (matchCallFuncToUse(use, "embedding_bag", 2)) {
+      embedding_bag_float_op = use.user;
+    }
+  }
```
what are the possible cases here? can we have stronger guarantees here? is the observer_out always going to be used by an embedding_bag op here?

supriyar committed:
Update on "[quant] Add Graph Mode Passes to quantize EmbeddingBag operators"


Summary:
This change adds preliminary support to quantize the EmbeddingBag operators. We currently support 4-bit and 8-bit quantization+packing of the weights.

To quantize these operators, specify the operator name in the `custom_op_name` field of the NoopObserver. Based on the op name (4bit or 8bit) we call the corresponding quantization functions.
Refer to the testplan for how to invoke the qconfig for the embedding_bag ops.

Future versions of this will support 4-bit and 2-bit qtensors with native support to observe and quantize it.

NB - This version assumes that the weights in the EmbeddingBag Module reside on the same device.

Test Plan:
python test/test_quantization.py TestQuantizeDynamicJitOps.test_embedding_bag

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22609342](https://our.internmc.facebook.com/intern/diff/D22609342)

[ghstack-poisoned]

supriyar committed:
Update on "[quant] Add Graph Mode Passes to quantize EmbeddingBag operators"


Summary:
This change adds preliminary support to quantize the EmbeddingBag operators. We currently support 4-bit and 8-bit quantization+packing of the weights.

To quantize these operators, specify the operator name in the `custom_op_name` field of the NoopObserver. Based on the op name (4bit or 8bit) we call the corresponding quantization functions.
Refer to the testplan for how to invoke the qconfig for the embedding_bag ops.

Future versions of this will support 4-bit and 2-bit qtensors with native support to observe and quantize it.

NB - This version assumes that the weights in the EmbeddingBag Module reside on the same device.

Test Plan:
python test/test_quantization.py TestQuantizeDynamicJitOps.test_embedding_bag

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22609342](https://our.internmc.facebook.com/intern/diff/D22609342)

[ghstack-poisoned]

supriyar committed:
Update on "[quant] Add Graph Mode Passes to quantize EmbeddingBag operators"


Summary:
This change adds preliminary support to quantize the EmbeddingBag operators. We currently support 4-bit and 8-bit quantization+packing of the weights.

To quantize these operators, specify the operator name in the `custom_op_name` field of the NoopObserver. Based on the op name (4bit or 8bit) we call the corresponding quantization functions.
Refer to the testplan for how to invoke the qconfig for the embedding_bag ops.

Future versions of this will support 4-bit and 2-bit qtensors with native support to observe and quantize it.

NB - This version assumes that the weights in the EmbeddingBag Module reside on the same device.

Test Plan:
python test/test_quantization.py TestQuantizeDynamicJitOps.test_embedding_bag

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22609342](https://our.internmc.facebook.com/intern/diff/D22609342)

[ghstack-poisoned]

jerryzh168 commented on torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp:
```diff
@@ -249,50 +379,46 @@ void insertQuantizationOps(
-      inputs.push_back(g->insertGetAttr(self, qparam_name));
+  // Temporary solution to quantize embedding_bag operators.
+  auto embedding_bag_name = getEmbeddingBagObsName(module, observer);
+  if (quant_type == QuantType::DYNAMIC && embedding_bag_name &&
```
can you merge this branch with the one lin L399 now?

supriyar replied:
I prefer keeping it separate since this is a special case and this code will be removed in the future. Wanted to make that a little more obvious :)

jerryzh168 replied:
might also be good to have a isEmbeddingBagOp function to be consistent with isFP16NoopObserver

jerryzh168 replied:
I see, sure. do you mean we plan to remove the swapping of input and weight in the future in the embedding bag module? I think this will be needed in graph mode if that does not change

jerryzh168 commented on torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp:
```diff
@@ -249,50 +379,46 @@ void insertQuantizationOps(
+    if (isWeight(module, observer_out)) {
+      auto op_name = embedding_bag_name.value();
+      Node* dequant = insertEmbeddingBagOps(observer, op_name);
+      observer_out->replaceAllUsesWith(original_val);
+      original_val->replaceAllUsesAfterNodeWith(dequant, dequant->output());
```
these two lines are the same as 417-419

jerryzh168 commented on torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp:
```diff
@@ -249,50 +379,46 @@ void insertQuantizationOps(
+          g, observer_out, dtype, at::Symbol::aten(quantize_func));
+    } else {
+      // For weight tensors we insert quant-dequant ops.
+      std::tie(quant, dequant) = insertDefaultObserverNodes(
```
btw do we need the `quant` value here?

jerryzh168 commented on torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp:
```diff
@@ -228,6 +245,119 @@ bool isNoopObserver(Value* observer) {
+
+// Insert quant and dequant nodes into the graph for both static and dynamic
+// quant.
+std::tuple<Node*, Node*> insertDefaultObserverNodes(
```
suggest to rename this to insertQuantDeQuant

jerryzh168 reviewed (approved):
Looks good! thanks. had a few more inline comments.

jerryzh168 commented on torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp:
```diff
@@ -218,6 +218,23 @@ Node* insertFP16CastOps(Graph* graph, Value* observer_out) {
 }
 
+// find the observer for Value `v` and return the name of the observer
+c10::optional<std::string> findObserverName(Value* v) {
```
btw we can check for types now I think

supriyar committed:
Update on "[quant] Add Graph Mode Passes to quantize EmbeddingBag operators"


Summary:
This change adds preliminary support to quantize the EmbeddingBag operators. We currently support 4-bit and 8-bit quantization+packing of the weights.

To quantize these operators, specify the operator name in the `custom_op_name` field of the NoopObserver. Based on the op name (4bit or 8bit) we call the corresponding quantization functions.
Refer to the testplan for how to invoke the qconfig for the embedding_bag ops.

Future versions of this will support 4-bit and 2-bit qtensors with native support to observe and quantize it.

NB - This version assumes that the weights in the EmbeddingBag Module reside on the same device.

Test Plan:
python test/test_quantization.py TestQuantizeDynamicJitOps.test_embedding_bag

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22609342](https://our.internmc.facebook.com/intern/diff/D22609342)

[ghstack-poisoned]
